File: OS/ThreadLocalStorage.java

class MyThread {
    private static ThreadLocal<Integer> threadData = ThreadLocal.withInitial(() -> 0);  // ThreadLocal is a class that provides thread-local variables. Each thread accessing the variable has its own, independently initialized copy of the variable.

    public void increment() {
        threadData.set(threadData.get() + 1);
        System.out.println("Thread " + Thread.currentThread().getId() + " Counter: " + threadData.get());
    }
}

public class ThreadLocalStorage {
    public static void main(String[] args) {
        MyThread myThread = new MyThread();

        Thread t1 = new Thread(myThread::increment);
        Thread t2 = new Thread(myThread::increment);

        t1.start();
        t2.start();
    }
}

File: OS/ThreadLocalStorage.cpp
#include <iostream>
#include <thread>

//thread_local int counter = 0;  // Each thread gets its own counter
static __thread int counter = 0;  // The gcc compiler provides the storage class keyword __thread for declaring TLS data

void increment() {
    counter++;  
    std::cout << "Thread ID: " << std::this_thread::get_id() 
              << " Counter: " << counter << std::endl;
}

int main() {
    std::thread t1(increment);
    std::thread t2(increment);

    t1.join();
    t2.join();
}
